syntax = "proto3";

service NameNodeService {
    rpc RegisterDataNode (DataNodeInfo) returns (RegistrationResponse);
    rpc GetBlockLocations (FileRequest) returns (BlockLocations);
    rpc CreateFile (FileMetadata) returns (FileResponse);
    rpc ListFiles (ListRequest) returns (ListResponse);
}

message DataNodeInfo {
    string host = 1;
    int32 port = 2;
}

message RegistrationResponse {
    bool success = 1;
    string message = 2;
}

message FileRequest {
    string filename = 1;
}

message BlockLocations {
    repeated BlockInfo blocks = 1;
}

message BlockInfo {
    string block_id = 1;
    repeated DataNodeInfo datanodes = 2;
}

message FileMetadata {
    string filename = 1;
    int64 size = 2;
    int32 replication_factor = 3;
}

message FileResponse {
    bool success = 1;
    repeated BlockInfo blocks = 2;
}

message ListRequest {
    string path = 1;
}

message ListResponse {
    repeated string files = 1;
}